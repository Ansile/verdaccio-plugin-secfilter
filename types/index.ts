import { Config } from '@verdaccio/types';
import { Range } from 'semver';

export type PackageBlockRule = { scope: string } | { package: string } | { package: string; version: string };
export interface CustomConfig extends Config {
  dateThreshold?: string | number;
  block?: Array<PackageBlockRule>;
}

export type ParsedBlockRule = Range[] | 'scope' | 'package' | undefined;
