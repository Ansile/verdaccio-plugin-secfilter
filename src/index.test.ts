import { Package, Version, Logger } from '@verdaccio/types';

import * as semver from 'semver';
import { ParsedBlockRule } from '../types';

import { filterBlockedVersions } from './index';

const exampleVersion: Version = {
  _id: '',
  main: '',
  name: '',
  readme: '',
  version: '',
} as any;

const examplePackage: Package = {
  'dist-tags': { latest: '3.0.0' },
  _attachments: {},
  _distfiles: {},
  _rev: '',
  _uplinks: {},
  name: '@babel/test',
  versions: {
    '1.0.0': exampleVersion,
    '3.0.0': exampleVersion,
  },
};


const noop = () => {};
const logger: Logger = {
  child: noop,
  debug: noop,
  error: noop,
  http: noop,
  warn: noop,
  info: noop,
  trace: noop,
};

describe('filters blocked packages', () => {
  test('filters when rules are for scope', () => {
    const block = new Map<string, ParsedBlockRule>([['@babel', 'scope']]);

    expect(filterBlockedVersions(examplePackage, block, logger)).toMatchSnapshot();
  });

  test('filters when rules are for package', () => {
    const block = new Map<string, ParsedBlockRule>([['@babel/test', 'package']]);

    expect(filterBlockedVersions(examplePackage, block, logger)).toMatchSnapshot();
  });

  test('filters when rules are for versions', () => {
    const block = new Map<string, ParsedBlockRule>([['@babel/test', { block: [new semver.Range('>1.0.0')] }]]);

    expect(filterBlockedVersions(examplePackage, block, logger)).toMatchSnapshot();
  });

  test('filters when multiple rules are for versions', () => {
    const block = new Map<string, ParsedBlockRule>([
      ['@babel/test', { block: [new semver.Range('>1.0.0'), new semver.Range('<=1.0.0')] }],
    ]);

    expect(filterBlockedVersions(examplePackage, block, logger)).toMatchSnapshot();
  });
});
