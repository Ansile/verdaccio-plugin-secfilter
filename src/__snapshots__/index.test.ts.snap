// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filters blocked packages filters when multiple rules are for versions 1`] = `
Object {
  "_attachments": Object {},
  "_distfiles": Object {},
  "_rev": "",
  "_uplinks": Object {},
  "dist-tags": Object {
    "latest": "3.0.0",
  },
  "name": "@babel/test",
  "readme": "
Some versions of package are blocked by rules: >1.0.0,<=1.0.0",
  "versions": Object {},
}
`;

exports[`filters blocked packages filters when rules are for package 1`] = `
Object {
  "_attachments": Object {},
  "_distfiles": Object {},
  "_rev": "",
  "_uplinks": Object {},
  "dist-tags": Object {
    "latest": "3.0.0",
  },
  "name": "@babel/test",
  "readme": "All package versions blocked by rule",
  "versions": Object {},
}
`;

exports[`filters blocked packages filters when rules are for scope 1`] = `
Object {
  "_attachments": Object {},
  "_distfiles": Object {},
  "_rev": "",
  "_uplinks": Object {},
  "dist-tags": Object {
    "latest": "3.0.0",
  },
  "name": "@babel/test",
  "readme": "All packages in scope @babel blocked by rule",
  "versions": Object {},
}
`;

exports[`filters blocked packages filters when rules are for versions 1`] = `
Object {
  "_attachments": Object {},
  "_distfiles": Object {},
  "_rev": "",
  "_uplinks": Object {},
  "dist-tags": Object {
    "latest": "3.0.0",
  },
  "name": "@babel/test",
  "readme": "
Some versions of package are blocked by rules: >1.0.0",
  "versions": Object {
    "1.0.0": Object {
      "_id": "",
      "main": "",
      "name": "",
      "readme": "",
      "version": "",
    },
  },
}
`;
